File: docs-examples/example-versions-java/src/test/java/io/micronaut/configuration/openapi/docs/versions/VersionedControllerTest.java
package io.micronaut.configuration.openapi.docs.versions;

import io.swagger.v3.oas.models.media.Schema;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class VersionedControllerTest {

    @Test
    void version1() {
        var openApi = OpenApiSpec.load("service-1.0.0-1.yml");

        var common = openApi.getPaths().get("/versioned/common").getPost();
        assertEquals("common", common.getOperationId());
        assertVersionParameter(common.getParameters().get(0));
        var hello = openApi.getPaths().get("/versioned/hello").getGet();
        assertEquals("helloV1", hello.getOperationId());
        assertVersionParameter(hello.getParameters().get(0));
        assertFalse(openApi.getPaths().get("/versioned/hello").readOperationsMap().containsKey(io.swagger.v3.oas.models.PathItem.HttpMethod.POST));
    }

    @Test
    void version2() {
        var openApi = OpenApiSpec.load("service-1.0.0-2.yml");

        var common = openApi.getPaths().get("/versioned/common").getPost();
        assertEquals("common", common.getOperationId());
        var hello = openApi.getPaths().get("/versioned/hello").getPost();
        assertEquals("helloV2", hello.getOperationId());
        assertVersionParameter(hello.getParameters().get(0));
        Schema<?> body = hello.getRequestBody().getContent().get("application/json").getSchema();
        String userDtoRef = body.getProperties().get("userDto").get$ref();
        assertTrue(userDtoRef.endsWith("UserDto"));
        Schema<?> userDto = openApi.getComponents().getSchemas().get(userDtoRef.substring("#/components/schemas/".length()));
        assertEquals(List.of("address"), userDto.getRequired());
        assertEquals(List.of("name", "age", "secondName", "address"), List.copyOf(userDto.getProperties().keySet()));
        assertFalse(openApi.getPaths().get("/versioned/hello").readOperationsMap().containsKey(io.swagger.v3.oas.models.PathItem.HttpMethod.GET));
    }

    private static void assertVersionParameter(io.swagger.v3.oas.models.parameters.Parameter parameter) {
        assertEquals("version", parameter.getName());
        assertEquals("query", parameter.getIn());
        assertEquals("API version", parameter.getDescription());
        assertEquals("string", parameter.getSchema().getType());
    }
}

File: docs-examples/example-versions-python/src/test/python/micronaut/configuration/openapi/docs/versions/VersionedControllerTest.py
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from . import OpenApiSpec

from io.swagger.v3.oas.models.PathItem import HttpMethod


def assert_version_parameter(parameter):
    assert parameter.getName() == "version"
    assert parameter.getIn() == "query"
    assert parameter.getDescription() == "API version"
    assert parameter.getSchema().getType() == "string"


@MicronautTest
class VersionedControllerTest:

    @Test
    def test_version_1(self):
        open_api = OpenApiSpec.load("service-1.0.0-1.yml")

        common = open_api.getPaths().get("/versioned/common").getPost()
        assert common.getOperationId() == "common"
        assert_version_parameter(common.getParameters().get(0))
        hello = open_api.getPaths().get("/versioned/hello").getGet()
        assert hello.getOperationId() == "helloV1"
        assert_version_parameter(hello.getParameters().get(0))
        assert not open_api.getPaths().get("/versioned/hello").readOperationsMap().containsKey(HttpMethod.POST)

    @Test
    def test_version_2(self):
        open_api = OpenApiSpec.load("service-1.0.0-2.yml")

        common = open_api.getPaths().get("/versioned/common").getPost()
        assert common.getOperationId() == "common"
        hello = open_api.getPaths().get("/versioned/hello").getPost()
        assert hello.getOperationId() == "helloV2"
        assert_version_parameter(hello.getParameters().get(0))
        body = hello.getRequestBody().getContent().get("application/json").getSchema()
        # `Schema.get$ref()` has no alias (`$` cannot appear in a Python identifier), so it is looked up by name
        user_dto_ref = getattr(body.getProperties().get("userDto"), "get$ref")()
        assert user_dto_ref.endswith("UserDto")
        user_dto = open_api.getComponents().getSchemas().get(user_dto_ref.removeprefix("#/components/schemas/"))
        assert "address" in list(user_dto.getRequired())
        assert list(user_dto.getProperties().keySet()) == ["name", "age", "secondName", "address"]
        assert not open_api.getPaths().get("/versioned/hello").readOperationsMap().containsKey(HttpMethod.GET)
